runtime/common/src/migrations/loans.rs

```diff
@@ -0,0 +1,168 @@
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::PoolNAV;
+use frame_support::{
+	pallet_prelude::{Decode, Encode, StorageVersion},
+	traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+	weights::Weight,
+};
+use pallet_loans::{pallet::Pallet as Loans, Config};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::TryRuntimeError;
+use sp_std::vec::Vec;
+
+const LOG_PREFIX: &str = "LoansMigrationToV4:";
+
+mod v3 {
+	use frame_support::{
+		pallet_prelude::{OptionQuery, ValueQuery},
+		storage_alias, Blake2_128Concat, BoundedVec,
+	};
+	pub use pallet_loans::entities::{
+		loans::v3::ActiveLoan, pricing::external::v3::CreatedLoan as CreatedLoanStruct,
+	};
+	use pallet_loans::Config;
+
+	use super::Loans;
+
+	pub type ActiveLoansVec<T> =
+		BoundedVec<(<T as Config>::LoanId, ActiveLoan<T>), <T as Config>::MaxActiveLoansPerPool>;
+
+	#[storage_alias]
+	pub type ActiveLoans<T: Config> = StorageMap<
+		Loans<T>,
+		Blake2_128Concat,
+		<T as Config>::PoolId,
+		ActiveLoansVec<T>,
+		ValueQuery,
+	>;
+
+	#[storage_alias]
+	pub type CreatedLoan<T: Config> = StorageDoubleMap<
+		Loans<T>,
+		Blake2_128Concat,
+		<T as Config>::PoolId,
+		Blake2_128Concat,
+		<T as Config>::LoanId,
+		CreatedLoanStruct<T>,
+		OptionQuery,
+	>;
+}
+
+pub struct AddWithLinearPricing<T>(sp_std::marker::PhantomData<T>);
+impl<T> OnRuntimeUpgrade for AddWithLinearPricing<T>
+where
+	T: Config,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+		assert_eq!(
+			Loans::<T>::on_chain_storage_version(),
+			StorageVersion::new(3)
+		);
+
+		let created_loans: u64 = v3::CreatedLoan::<T>::iter_keys().count().saturated_into();
+		let active_loans: u64 = v3::ActiveLoans::<T>::iter_values()
+			.map(|v| v.len())
+			.sum::<usize>()
+			.saturated_into();
+
+		log::info!("{LOG_PREFIX} Pre checks done!");
+
+		Ok((created_loans, active_loans).encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		let mut weight = T::DbWeight::get().reads(1);
+		if Loans::<T>::on_chain_storage_version() == StorageVersion::new(3) {
+			log::info!("{LOG_PREFIX} Starting migration v3 -> v4");
+
+			pallet_loans::CreatedLoan::<T>::translate::<v3::CreatedLoanStruct<T>, _>(
+				|_, _, loan| {
+					weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+					Some(loan.migrate(true))
+				},
+			);
+
+			let mut changed_pools = Vec::new();
+			pallet_loans::ActiveLoans::<T>::translate::<v3::ActiveLoansVec<T>, _>(
+				|pool_id, active_loans| {
+					changed_pools.push(pool_id);
+					weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+					Some(
+						active_loans
+							.into_iter()
+							.map(|(loan_id, active_loan)| (loan_id, active_loan.migrate(true)))
```

Comment:
Are we sure we want all loans `with_linear_accrual == true`? cc @mustermeiszer

Reply:
Yes please double check this. The implementation is based on [this comment](https://kflabs.slack.com/archives/C04GJQAM9P0/p1715871507753699?thread_ts=1715849878.667849&cid=C04GJQAM9P0) from @hieronx.

Reply:
Just confirmed again with Anemoy, yes they want linear pricing to be on for all existing assets.